refactor and add tests

rgalonso opened the pull request:
This contains some additional minor refactoring and adds another unittest file.

## refactor-related changes
The following changes were made further isolate the VCS-specific code from that in `main`. 
- added abstract `Client` base class
  - this also supports `mypy`-related changes, discussed below
- parse `MANUAL_COMMIT_REF` and `MANUAL_BASE_REF` environment variables within *`Client` so that diff URL can be VCS-specific
  - for `LocalClient`, there is no URL, but parsing these variables allows it to call `git diff` with the proper arguments
- `Client:get_issue_url()` returns _just_ the URL string (when applicable), rather than prefixing it with "Issue URL: ". This allows the function to be used more generically.

## more thorough testing
In addition, per [this discussion](https://github.com/alstr/todo-to-issue-action/pull/217#issuecomment-2441838974) in PR #217, I added an additional test file which uses `mypy` to look for additional errors that weren't being caught before. `mypy` is really intended for layering static typing on top of Python. That's a nice goal, but I've limited the changes here to most just focus on having the additional testing without being concerned about the typing.

## other minor items
- added devcontainer configuration
  - to support local development
- added additional (non-defualt) Dockerfile target
  - to support local testing and upcoming GitLab integration
- minor changes to requirements/-dev.txt and GitHub workflow to support above changes

alstr commented:
This is awesome. Thanks very much for the effort. I'll take a look over it shortly!

rgalonso commented:
> This is awesome. Thanks very much for the effort. I'll take a look over it shortly!

In particular, please pay attention to how this is working for GitHub Actions. Since I'm focused on GitLab, I'm not testing the GitHub side of it all. I'm not even sure how to! So I'm definitely relying on you and any existing regression tests to make sure that side of it is still working properly. (You can probably take that as a general note as I continue to submit new PRs.) Thanks again!

alstr commented:
All looks good, thanks again. About the update to the Dockerfile; what's the purpose of the additional stage? Just wondering if it'll slow things down.

rgalonso commented:
That's a fair concern, but with modern Docker it's not an issue. It used to be that it would build all stages, but these days it's smart enough to skip the new stage I introduced because it can see that the final (unnamed) stage doesn't rely on it.

As to why I introduced it, it's needed for GitLab CI/CD. Independent of that, I've also found it useful for local testing prior to a push.

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________
From: Alastair Mooney ***@***.***>
Sent: Monday, November 4, 2024 7:05:36 AM
To: alstr/todo-to-issue-action ***@***.***>
Cc: Robert Alonso ***@***.***>; Author ***@***.***>
Subject: Re: [alstr/todo-to-issue-action] refactor and add tests (PR #222)

Caution: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe.


All looks good, thanks again. About the update to the Dockerfile; what's the purpose of the additional stage? Just wondering if it'll slow things down.

—
Reply to this email directly, view it on GitHub<https://github.com/alstr/todo-to-issue-action/pull/222#issuecomment-2454540832>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEFHTTIK4HW7MNGRGADUH5LZ65PIZAVCNFSM6AAAAABQYDFAXGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDINJUGU2DAOBTGI>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

alstr commented:
Thanks for that.

I've merged the PR, adding a couple of other bits in the meantime:

- Added server URL to `GitHubClient` (#218)
- Importing `TodoParser` from `main.py` was failing, so I updated it to import directly

I notice the new tests are failing and that this is expected:

https://github.com/alstr/todo-to-issue-action/actions/runs/11681623509/job/32527038946

Should I update the GitHub Actions workflow run to exclude these tests in that case?

rgalonso commented:
> Thanks for that.
> 
> I've merged the PR, adding a couple of other bits in the meantime:
> 
> * Added server URL to `GitHubClient` ([Fix link generation for GitHub Enterprise by using `GITHUB_SERVER_URL` #218](https://github.com/alstr/todo-to-issue-action/pull/218))
> * Importing `TodoParser` from `main.py` was failing, so I updated it to import directly
> * `Client`, `GitHubClient`, `LocalClient`, `LineStatus`, `TodoParser` were undefined in `main.py` so I have also imported them manually. At first I thought this was to be expected looking at `test_todo_parser.py`, but having looked into it further I don't think these errors were expected.

I think the reason the `TodoParser` import was failing is that you may have had a bad merge conflict when you [merged](https://github.com/alstr/todo-to-issue-action/pull/218/commits/37b15115baef6cc67c6531a065d96db7b0ede06f) `master` to `ghe-support` prior to the PR #218 merge. Was that a conflict you had to fix manually? Or did the automerge mess up? It looks like the changes from `master` to `main.py` (in particular those of commit 662435d7fc77b574ff7de82f8cc5b8126d0e6c22) _mostly_ made it into the merge commit, but not entirely. Most of the `main.py` changes of 662435d7fc77b574ff7de82f8cc5b8126d0e6c22 were deletions, but there were a few critical additions (below), and those are *not* found in the merge commit.
```
+from TodoParser import *
+from LineStatus import *
+from GitHubClient import *
```
Yes, the [check](https://github.com/alstr/todo-to-issue-action/pull/218/checks) failed, but the reason was because `main.py` didn't have the above lines, so `from main import TodoParser` couldn't possibly work. Had the `main.py` merge gone correctly, there wouldn't have been an issue.

Thankfully, it looks like you caught it and fixed it in efb8b9830dd1eb59314ac2e886af7b917fad7d21, so I'm probably not telling you anything you don't know. But in case you didn't realize this is what happened, I wanted to highlight it just in case there's something about your local dev environment or process that caused this error or made you overlook the merge conflict (if there was one even reported).
